src/document.rs

```diff
@@ -129,172 +172,447 @@ pub struct Prescription<'a> {
     pub document: Document<'a>,
 }
 
-mod helpers {
-    use super::unicode_segmentation::UnicodeSegmentation;
-    use super::NodeType;
-
-    /// Sequenceable node types are those which can be used as a pattern in the prescription
-    /// that can match against an arbitrary sequence of nodes of the same type in a document.
-    pub fn node_type_is_sequencable(node_type: &NodeType) -> bool {
-        match *node_type {
-            NodeType::CMarkNodeParagraph
-            | NodeType::CMarkNodeHeading
-            | NodeType::CMarkNodeItem
-            | NodeType::CMarkNodeList
-            | NodeType::CMarkNodeCodeBlock => true,
-            _ => false,
-        }
+/// A circumstantial node is one that may or may not appear in the document due to its presence
+/// only being necessary if an optional child node is present.
+pub fn node_type_is_circumstantial(node: &Node) -> HowserResult<bool> {
+    let node_type = node.capabilities
+        .get
+        .as_ref()
+        .ok_or(HowserError::CapabilityError)?
+        .get_type()?;
+    match node_type {
+        NodeType::CMarkNodeList => Ok(true),
+        _ => Ok(false),
     }
+}
+
+pub fn reverse(s: &String) -> String {
+    let iter = UnicodeSegmentation::graphemes(s.as_str(), true);
+    iter.rev().collect::<String>()
+}
 
-    /// A circumstantial node is one that may or may not appear in the document due to its presence
-    /// only being necessary if an optional child node is present.
-    pub fn node_type_is_circumstantial(node_type: &NodeType) -> bool {
-        match *node_type {
-            NodeType::CMarkNodeList => true,
-            _ => false,
+/// Extract and transform the block-level Rx prompts from the document content into
+/// contextualized metadata on the match types of those elements.
+fn process_child_block_elements(parent: &Node, document: &Document) -> HowserResult<()> {
```

Comment:
I may be misreading, but it looks like this fn is doing two things: walking the document tree, and then altering it on the basis of the nodes it encounters in that traversal (with lookahead). If htat's correct, the code could likely be made simpler by splitting the two up: have an iterator to do the traversal and provide lookahead, then pass the iterator into a fn which looks for the right pattern and changes things.

Reply:
(This is the visitor pattern, kind of)

Reply:
That's an interesting thought. I'm not sure it will be easy to apply in this case, but I'd like to chat about the concept with you some more.